dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/orgunitprofile/impl/DefaultOrgUnitProfileService.java

```diff
@@ -0,0 +1,434 @@
+import org.hisp.dhis.organisationunit.OrganisationUnitGroupService;
+import org.hisp.dhis.organisationunit.OrganisationUnitGroupSet;
+import org.hisp.dhis.orgunitprofile.OrgUnitInfo;
+import org.hisp.dhis.orgunitprofile.OrgUnitProfile;
+import org.hisp.dhis.orgunitprofile.OrgUnitProfileData;
+import org.hisp.dhis.orgunitprofile.OrgUnitProfileService;
+import org.hisp.dhis.orgunitprofile.ProfileItem;
+import org.hisp.dhis.period.Period;
+import org.hisp.dhis.period.PeriodType;
+import org.hisp.dhis.period.RelativePeriodEnum;
+import org.hisp.dhis.period.RelativePeriods;
+import org.hisp.dhis.program.ProgramIndicator;
+import org.locationtech.jts.geom.Point;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+
+@Slf4j
+@Service
+public class DefaultOrgUnitProfileService
+    implements OrgUnitProfileService
+{
+    private static final List<Class<? extends IdentifiableObject>> DATA_ITEM_CLASSES = ImmutableList
+        .<Class<? extends IdentifiableObject>> builder()
+        .add( DataElement.class ).add( Indicator.class ).add( DataSet.class ).add( ProgramIndicator.class )
+        .build();
+
+    private KeyJsonValueService dataStore;
+
+    private IdentifiableObjectManager idObjectManager;
+
+    private AnalyticsService analyticsService;
+
+    private OrganisationUnitGroupService groupService;
+
+    private ObjectMapper jsonMapper;
+
+    public DefaultOrgUnitProfileService( KeyJsonValueService dataStore,
+        IdentifiableObjectManager idObjectManager, AnalyticsService analyticsService,
+        OrganisationUnitGroupService groupService, ObjectMapper jsonMapper )
+    {
+        this.dataStore = dataStore;
+        this.idObjectManager = idObjectManager;
+        this.analyticsService = analyticsService;
+        this.jsonMapper = jsonMapper;
+        this.groupService = groupService;
+
+        this.dataStore.addProtection(
+            new KeyJsonNamespaceProtection( ORG_UNIT_PROFILE_NAMESPACE, KeyJsonNamespaceProtection.ProtectionType.NONE,
+                KeyJsonNamespaceProtection.ProtectionType.RESTRICTED, false, "ALL", ORG_UNIT_PROFILE_AUTHORITY ) );
+    }
+
+    @Override
+    @Transactional
+    public void saveOrgUnitProfile( OrgUnitProfile profile )
+    {
+        KeyJsonValue keyJsonValue = new KeyJsonValue( ORG_UNIT_PROFILE_KEY, ORG_UNIT_PROFILE_NAMESPACE );
+
+        try
+        {
+            keyJsonValue.setValue( jsonMapper.writeValueAsString( profile ) );
+            dataStore.saveOrUpdateKeyJsonValue( keyJsonValue );
+        }
+        catch ( JsonProcessingException e )
+        {
+            log.error( DebugUtils.getStackTrace( e ) );
+            throw new IllegalArgumentException( "Can't serialize OrgUnitProfile " + profile );
```

Comment:
Use `String.format` instead of string concatenation.

Reply:
fixed